pkg/apis/vpc/v1alpha1/subnet_types.go

```diff
@@ -35,9 +39,17 @@ type SubnetSpec struct {
 	// +kubebuilder:validation:MaxItems=2
 	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
 	IPAddresses []string `json:"ipAddresses,omitempty"`
-	// DHCPConfig DHCP configuration.
-	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
-	DHCPConfig DHCPConfig `json:"DHCPConfig,omitempty"`
+
+	// DHCP mode of a SubnetSet cannot switch from DHCPDeactivated to DHCPServer or DHCPRelay.
+	// If subnetDHCPConfig is not set, the DHCP mode is DHCPDeactivated by default.
+	// In order to enforce this rule, two XValidation rules are defined.
+	// The rule in SubnetSetSpec prevents the condition that subnetDHCPConfig is not set in
+	// old SubnetSetSpec while the new SubnetSetSpec specifies a field other than DHCPDeactivated.
+	// The rule in SubnetDHCPConfig prevents the mode changing from empty or
+	// DHCPDeactivated to DHCPServer or DHCPRelay.
+
+	// DHCP configuration for Subnet.
+	SubnetDHCPConfig SubnetDHCPConfig `json:"subnetDHCPConfig,omitempty"`
 }
 
 // SubnetStatus defines the observed state of Subnet.
```

Comment:
not related to this change, but networkAddress, gatewayAddress, dhcpServerAddress should be string instead of []string.
{
                "static_ip_pool_path": "/orgs/org-1/projects/project-1/vpcs/vpc-1/subnets/subnet-1/ip-pools/static-ipv4-default",
                "static_ip_pool_range": "11.0.0.171-11.0.0.190",
                "gateway_address": "11.0.0.129/26",
                "network_address": "11.0.0.128/26",
                "dhcp_server_address": "11.0.0.130/26",
                "dhcp_ranges": "11.0.0.131-11.0.0.170",
                "ip_address_type": "IPV4"
            }

Reply:
Thanks for pointing out, will improve this part in another PR.